### PR TITLE
Add health check

### PR DESCRIPTION
**Pull request by hasufell**
# Issue Number

ADP-496

## Details

- [x] Verify url and SMASH health on app start
- [x] Verify url and SMASH health on settings change
- [x] Expose SMASH health check proxy endpoints

### CODE REVIEW

File: lib/shelley/src/Cardano/Wallet/Shelley.hs [hunk 1]
```diff
@@ -351,6 +360,22 @@ serveWallet
                 $ \db@DBLayer{..} -> do
 
             forM_ settings $ atomically . putSettings
+            dbSettings <- atomically readSettings
+            -- initially verity smash server
+            case poolMetadataSource dbSettings of
```
Comment (hasufell) [inline lib/shelley/src/Cardano/Wallet/Shelley.hs]: This is more or less on application start. We crash the app if the server isn't reachable.

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 2]
```diff
@@ -688,12 +693,24 @@ monitorMetadata
 monitorMetadata tr gp DBLayer{..} = do
     settings <- atomically readSettings
     manager <- newManager defaultManagerSettings
+
+    case poolMetadataSource settings of
```
Comment (hasufell) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: This usually triggers when something changes the smash setting after the server started. If the server isn't healthy, we crash the worker thread.

The problem is... the caller doesn't really know, but they should! I don't think it's enough to emit warnings here, because subsequent fetch attempts will just fail forever. We probably need an API to query the status of the metadata worker.

**Review by rvl — COMMENTED** [review]
It's not clear from the PR description what this health check stuff is supposed to do. From all the discussion on that jira ticket, what did you end up deciding?

Can I suggest that a more user friendly approach would be to not crash when the status check fails? Perhaps log it at warning level and carry on. The smash server may be intermittently unreachable, but that's no reason to prevent starting the wallet.

If Daedalus want to prevent users from setting an incorrect smash url, and go back to the previous value if the new one was wrong, then perhaps provide an endpoint where they can do the status check _before_ changing the setting.

For example, a new endpoint `GET /v2/network/information/smash?url=https://smash.awstest.iohkdev.io`. (If the `url` query param is missing then it health checks the currently configured smash server.)

**Comment by hasufell** [comment]
@rvl What about just requiring deadalus to do the health check on their end? It isn't complicated and they have more ways to react to incorrect input than we have. I think it's indeed better the wallet just logs this.

**Comment by rvl** [comment]
It's possible. All Daedalus would need to know is whether a given smash url works - yes/no.
They may prefer better separation of concerns -- or may not care.
We can also basically proxy the smash health check, which is easy enough.

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 3]
```diff
@@ -740,6 +740,21 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
...
+        FetchSMASH uri -> do
+            let loop = do
+                    r <- healthCheck (Just trFetch) (unSmashServer uri) manager
+                    case r of
+                        (Just health)
+                            | isHealthy health -> pure ()
+                        _ -> do
+                            let ms = (* 1_000_000)
+                            let sleepTime = ms 60
```
Comment (hasufell) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: 60s is maybe a bit excessive?

**Comment by hasufell** [comment]
Should be all set

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 4]
```diff
@@ -2155,3 +2158,48 @@ type instance PostTransactionFeeDataT (n :: NetworkDiscriminant) =
...
+            "unavailable" -> pure Unavailable
+            "unreachable" -> pure Unreachable
+            "no_smash_configured" -> pure NoSMASH
+            e -> fail ("Unexpeced value: " <> e)
+
+instance ToJSON ApiHealthCheck where
+    toJSON Available = String "available"
+    toJSON Unavailable = String "unavailable"
+    toJSON Unreachable = String "unreachable"
+    toJSON NoSMASH = String "no_smash_configured"
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Types.hs]: Both instances would be better off derived using generics. We already have all the necessary plumbing for this. Please rewrite this using generic JSON. You'll need to change `NoSMASH` to `NoSmashConfigured`.

File: lib/core/src/Cardano/Wallet/Api/Types.hs [hunk 5]
```diff
@@ -2155,3 +2158,48 @@ type instance PostTransactionFeeDataT (n :: NetworkDiscriminant) =
...
+
+
+{-------------------------------------------------------------------------------
+                         SMASH interfacing types
+-------------------------------------------------------------------------------}
+
+-- | Parses the SMASH HealthCheck type from the SMASH API.
+data HealthCheck = HealthCheck
+    { status :: T.Text
+    , version :: T.Text
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Types.hs]: Do not use qualified import for `Text`, just `Text` is fine as per our coding style guide.

File: lib/shelley/src/Cardano/Wallet/Shelley.hs [hunk 6]
```diff
@@ -427,13 +431,19 @@ instance ToText ApplicationLog where
...
+            , T.pack $ uriToString id uri ""
+            ]
 
 instance HasPrivacyAnnotation ApplicationLog
 instance HasSeverityAnnotation ApplicationLog where
     getSeverityAnnotation = \case
         MsgStarting _ -> Info
         MsgNetworkName _ -> Info
         MsgServerStartupError _ -> Alert
+        MsgFailedConnectSMASH _ -> Error
```
Comment (KtorZ) [inline lib/shelley/src/Cardano/Wallet/Shelley.hs]: I don't think this deserve anything higher than `Warning`. A `Notice` will be fine. It's nothing wrong with the wallet application per se, and it won't prevent the wallet from operating other functionalities correctly.

File: lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs [hunk 7]
```diff
@@ -427,6 +434,23 @@ server byron icarus shelley spl ntp =
...
             = Handler $ fmap ApiT $ liftIO $ getSettings spl
 
+    smash :: Server SMASH
+    smash = getCurrentSmashHealth :<|> getUriSmashHealth
+      where
+        getHealth smashServer = liftIO $ do
+            manager <- newManager defaultManagerSettings
+            health <- healthCheck Nothing (unSmashServer smashServer) manager
+            pure $ maybe Unreachable
+                (\h -> if isHealthy h then Available else Unavailable) health
```
Comment (KtorZ) [inline lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs]: Why not have `healthCheck` returns the sum-type instead of Boolean which require conversion at this stage :thinking:  ?

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 8]
```diff
@@ -740,40 +742,55 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
...
+                    | rsIterNumber < maxRetries = pure (not b)
+                    | otherwise = pure False
+
+                ms = (* 1_000_000)
+                baseSleepTime = ms 15
+
+            retrying (constantDelay baseSleepTime) retryCheck checkHealth
+
+        -- no smash configured, equivalent to healthy
+        _ -> pure True
```
Comment (KtorZ) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: That's quite error prone and one more reason to have this function deals directly with the `HealthCheck` sum-type at this level. This really is `NoSmashConfigured`.

File: specifications/api/swagger.yaml [hunk 9]
```diff
@@ -1746,6 +1746,22 @@ components:
...
+
+    ApiHealthCheck: &ApiHealthCheck
+      type: string
+      enum: ["available", "unavailable", "unreachable", "no_smash_configured"]
+      description: |
+        The status of the SMASH server. Possible values are:
+          - available -> SMASH server is awaiting your requests
+          - unavailable -> SMASH server is running, but currently unavailable, try again in a short time
+          - unreachable -> SMASH server could not be reached or didn't return a health status
+          - no_smash_configured -> SMASH is currently not configured, adjust the Settings first
```
Comment (KtorZ) [inline specifications/api/swagger.yaml]: I'd suggest using a markdown table here, it reads better in the end documentation. See for instance the field `withdrawal` in: https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/postTransactionFee

File: lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs (hunk 7 above)
Comment (hasufell) [reply lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs]: Because `Available` and `Unavailable` is only a subset of the sum-type. `isHealthy` has a smaller domain and only encodes the interpretation of the `status` values returned from SMASH.

File: lib/core/src/Cardano/Wallet/Api/Types.hs (hunk 4 above)
Comment (hasufell) [reply lib/core/src/Cardano/Wallet/Api/Types.hs]: We have to use `UntaggedValue :: SumEncoding` for that and if someone adds a non-nullary constructor later, the json encoding is broken and whether our tests catch it depends on the Arbitrary instances.

```
  | UntaggedValue
    -- ^ Constructor names won't be encoded. Instead only the contents of the
    -- constructor will be encoded as if the type had a single constructor. JSON
    -- encodings have to be disjoint for decoding to work properly.
    --
    -- When decoding, constructors are tried in the order of definition. If some
    -- encodings overlap, the first one defined will succeed.
    --
    -- /Note:/ Nullary constructors are encoded as strings (using
    -- 'constructorTagModifier'). Having a nullary constructor alongside a
    -- single field constructor that encodes to a string leads to ambiguity.
    --
    -- /Note:/ Only the last error is kept when decoding, so in the case of
    -- malformed JSON, only an error for the last constructor will be reported.
```
I think this encoding is generally pretty fishy.

File: lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs (hunk 7 above)
Comment (hasufell) [reply lib/shelley/src/Cardano/Wallet/Shelley/Api/Server.hs]: I'll rework the types a bit to make it less ambiguous.

**Comment by hasufell** [comment]
smash_health

File: lib/core/src/Cardano/Pool/Metadata.hs [hunk 10]
```diff
@@ -59,15 +63,17 @@ import Control.Tracer
...
     , eitherDecodeStrict
     , fieldLabelModifier
     , genericParseJSON
     , genericToJSON
+    , parseJSON
+    , toJSON
     )
 import Data.Bifunctor
-    ( first )
+    ( Bifunctor (first) )
```
Comment (rvl) [inline lib/core/src/Cardano/Pool/Metadata.hs]: Is this change necessary?

File: lib/core/src/Cardano/Pool/Metadata.hs [hunk 11]
```diff
@@ -116,6 +122,10 @@ metadaFetchEp pid (StakePoolMetadataHash bytes)
     hashStr = T.unpack $ T.decodeUtf8 $ convertToBase Base16 bytes
     pidStr  = T.unpack $ toText pid
 
+-- TODO: use SMASH servant types
```
Comment (rvl) [inline lib/core/src/Cardano/Pool/Metadata.hs:129]: Actually, we don't necessarily want to couple so tightly with the IOHK Haskell implementation of SMASH. It may happen that other SMASH servers are used instead.

There should be an API spec for it somewhere, and we should just work in reference to that.

Likewise, in `README.md`, I see a table containing SMASH software versions. That is not especially relevant to users, because they are not the ones operating the SMASH server(s).

File: lib/core/src/Cardano/Pool/Metadata.hs [hunk 12]
```diff
@@ -170,20 +180,24 @@ registryUrlBuilder baseUrl pid _ hash =
...
-    :: Tracer IO StakePoolMetadataFetchLog
+trMaybe
+    :: Maybe (Tracer IO StakePoolMetadataFetchLog)
+    -> StakePoolMetadataFetchLog
+    -> IO ()
+trMaybe mtr m = forM_ mtr $ \tr -> traceWith tr m
+
+-- | A smash GET request that reads the result at once into memory.
+smashRequest
+    :: Maybe (Tracer IO StakePoolMetadataFetchLog)
```
Comment (rvl) [inline lib/core/src/Cardano/Pool/Metadata.hs]: You will never need optional tracers like this.

There is `nullTracer :: Tracer m a` for that, or simply `mempty`.

File: lib/core/src/Cardano/Pool/Metadata.hs [hunk 13]
```diff
@@ -199,6 +213,53 @@ fetchDelistedPools tr uri manager = runExceptTLog $ do
...
+    -> URI
+    -> Manager
+    -> IO (Maybe HealthStatusSMASH)
+healthCheck mtr uri manager = runExceptTLog $ do
+    pl <- smashRequest mtr
+        (uri { uriPath = "/" <> healthCheckEP , uriQuery = "", uriFragment = "" })
+        manager
+    except . eitherDecodeStrict @HealthStatusSMASH $ pl
+  where
+    runExceptTLog
```
Comment (rvl) [inline lib/core/src/Cardano/Pool/Metadata.hs:233]: Here's a tip.
You do not need trace message constructors for both Success and Failure.
Merge the log message constructors to `MsgFetchHealthCheckResult (Either FailureType HealthCheckSuccessType)`.
All your logging code will become simpler like that.

File: lib/core/src/Cardano/Pool/Metadata.hs [hunk 14]
```diff
@@ -199,6 +213,53 @@ fetchDelistedPools tr uri manager = runExceptTLog $ do
...
+        :: ExceptT String IO HealthStatusSMASH
+        -> IO (Maybe HealthStatusSMASH)
+    runExceptTLog action = runExceptT action >>= \case
+        Left msg ->
+            Nothing <$ trMaybe mtr (MsgFetchHealthCheckFailure msg)
+
+        Right health
+            | isHealthyStatus health -> do
+                trMaybe mtr (MsgFetchHealthCheckFailure
+                    ("Server reports unhealthy status: " <> T.unpack (status health)))
```
Comment (rvl) [inline lib/core/src/Cardano/Pool/Metadata.hs]: It's better to do log message formatting in the ToText StakePoolMetadataFetchLog instance.
Then the program logic is not obscured by logging.

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 15]
```diff
@@ -740,40 +743,56 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
...
-                    fetchDelistedPools trFetch (toDelistedPoolsURI uri) manager
-            tid <- forkFinally
-                (gcDelistedPools gcStatus tr db getDelistedPools)
-                onExit
-            flip finally (killThread tid) $
-                loop (fetchThem $ fetcher [registryUrlBuilder uri])
+    health <- case poolMetadataSource settings of
+        FetchSMASH uri -> do
+            let checkHealth _ = do
+                    r <- healthCheck (Just trFetch) (unSmashServer uri) manager
```
Comment (rvl) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: Is it really necessary to run a health check before fetching data?
If the server is healthy it will return data. If it's not healthy it will return some kind of error from the fetch.

**Review by rvl — DISMISSED** [review]

File: lib/core/src/Cardano/Pool/Metadata.hs (hunk 11 above)
Comment (hasufell) [reply lib/core/src/Cardano/Pool/Metadata.hs:129]: I disagree. The entire point of servant is to be able to consume its types.

> It may happen that other SMASH servers are used instead.

That sounds too theoretical to base implementation decisions on it.

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs (hunk 15 above)
Comment (hasufell) [reply lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: > Is it really necessary to run a health check before fetching data?

Yes, otherwise you're fetching data from an unknown server that may respond with whatever data. SMASH is configurable and we don't control where it points to.

That's the whole point of this ticket: https://jira.iohk.io/browse/ADP-496

File: lib/core/src/Cardano/Pool/Metadata.hs (hunk 11 above)
Comment (hasufell) [reply lib/core/src/Cardano/Pool/Metadata.hs:129]: > Likewise, in README.md, I see a table containing SMASH software versions. That is not especially relevant to users, because they are not the ones operating the SMASH server(s).

Feel free to discuss that with @KtorZ who requested this change.

File: lib/core/src/Cardano/Pool/Metadata.hs (hunk 13 above)
Comment (hasufell) [reply lib/core/src/Cardano/Pool/Metadata.hs:233]: Failure here is HTTP failure, success is just "we got an answer" and doesn't mean the server is healthy. I'd like to keep those separate, because they currently don't have the same severity either.

Comment (hasufell) [reply lib/core/src/Cardano/Pool/Metadata.hs:233]: But you are right in the sense that the response object shouldn't have information that diverges from the HTTP status code it returns.

File: lib/core/src/Cardano/Pool/Metadata.hs (hunk 11 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Pool/Metadata.hs:129]: > That sounds too theoretical to base implementation decisions on it.

It is already happening as we speak. Adapools has its own SMASH server. This is why I've also repeated in the past that we want to avoid re-using SMASH's internal types. We do integrate against the interface (a.k.a OpenAPI), not the implementation. 
Now, we'll probably end up using servant types _anyway_ on our end, so it could be interesting to define them in a library that is self contained, but this should not create any strong coupling with SMASH and SMASH development. 

> Likewise, in README.md, I see a table containing SMASH software versions. That is not especially relevant to users, because they are not the ones operating the SMASH server(s).

@rvl I've asked Julian to add that because it does actually concern users. Users are going to point cardano-wallet to SMASH servers and they want to know with which version of SMASH it works. And by version here, I mean the version of the corresponding OpenAPI specification, not the version of SMASH-the-haskell-executable (although arguably, major and minor should move in locksteps for these).

Comment (hasufell) [reply lib/core/src/Cardano/Pool/Metadata.hs:129]: > Now, we'll probably end up using servant types anyway on our end, so it could be interesting to define them in a library that is self contained, but this should not create any strong coupling with SMASH and SMASH development.

SMASH already exposes the servant types separately: https://github.com/input-output-hk/smash/tree/master/smash-servant-types

The servant types must be checked against the OpenAPI spec (or generated from it). Then anyone else can still use whatever language they want and we can still use the servant types, because these are isomorphic. The spec part is still TODO on SMASHs end afais, but once we have it, there's no reason not to use the types.

File: lib/core/src/Cardano/Pool/Metadata.hs [hunk 16]
```diff
@@ -199,6 +207,47 @@ fetchDelistedPools tr uri manager = runExceptTLog $ do
...
+    | T.toLower status /= "ok" = True
+    | otherwise = False
+
+fetchDelistedPools
+    :: Tracer IO StakePoolMetadataFetchLog
+    -> URI
+    -> Manager
+    -> IO (Maybe [PoolId])
+fetchDelistedPools tr uri manager = runExceptTLog $ do
+    pl <- smashRequest tr uri manager
```
Comment (hasufell) [inline lib/core/src/Cardano/Pool/Metadata.hs]: TODO: 

- [x] this is broken and needs to expand to the delisted endpoint

File: lib/core/src/Cardano/Pool/Metadata.hs (hunk 13 above)
Comment (rvl) [reply lib/core/src/Cardano/Pool/Metadata.hs:233]: Whatever "result" type you choose, it can be pattern matched in the `getSeverityAnnotation` definition.

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs (hunk 15 above)
Comment (rvl) [reply lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: Actually the ticket description says "Daedalus has no way to know whether such a [SMASH] url points to a valid server."
So this change should just provide a way for Daedalus to check that a SMASH URL is correct before applying the setting.

There is no real need to call the health check API before fetching data. Why does everything become more complicated when a health check function is added?

Comment (hasufell) [reply lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: > So this change should just provide a way for Daedalus to check that a SMASH URL is correct before applying the setting.

The ticket says that we want to know whether it points to a valid server, not whether the URL is valid. That's not the same thing. We are dealing with user input here. We don't want to start fetching stuff randomly and then show errors. We want to know ahead of time whether a URL points to an actual SMASH server and that server is ready to accept requests. And if it isn't, daedalus can reject the setting. If we just apply it and then error out when actual metadata is fetched, then that's poor user experience.

File: specifications/api/swagger.yaml [hunk 17]
```diff
@@ -1700,6 +1728,25 @@ components:
...
+      enum: ["available", "unavailable", "unreachable", "no_smash_configured"]
+      description: |
+        The status of the SMASH server. Possible values are:
+
+        status                  | description
+        ---                     | ---
+        `"available"`           | server is awaiting your requests
+        `"unavailable"`         | server is running, but currently unavailable, try again in a short time
+        `"unreachable"`         | server could not be reached or didn't return a health status
+        `"no_smash_configured"` | SMASH is currently not configured, adjust the Settings first
```
Comment (KtorZ) [inline specifications/api/swagger.yaml:1725]: The health check status as a raw string feels a bit rough and prevent any further extension. What about wrapping this one in a `{ "health": ... }` or `{ "smash": ... }` or `{ "status": ... }` singleton..

File: specifications/api/swagger.yaml [hunk 18]
```diff
@@ -389,6 +390,33 @@ x-ScriptValue: &ScriptValue
 
 x-script: &script
   <<: *ScriptValue
+  description: |
```
Comment (KtorZ) [inline specifications/api/swagger.yaml]: What exactly is this one doing here ? Sounds like a bad merge resolution.

File: specifications/api/swagger.yaml [hunk 19]
```diff
@@ -341,6 +341,7 @@ x-ScriptValue: &ScriptValue
       type: string
       format: bech32
       pattern: "^(script_vkh)1[0-9a-z]*$"
+      example: script_vkh18srsxr3khll7vl3w9mqfu55n6wzxxlxj7qzr2mhnyreluzt36ms
```
Comment (KtorZ) [inline specifications/api/swagger.yaml]: ?

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 20]
```diff
@@ -740,40 +743,56 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
...
-                onExit
-            flip finally (killThread tid) $
-                loop (fetchThem $ fetcher [registryUrlBuilder uri])
+    health <- case poolMetadataSource settings of
+        FetchSMASH uri -> do
+            let checkHealth _ = do
+                    r <- healthCheck trFetch (unSmashServer uri) manager
+                    case r of
+                        (Just health)
+                            | isHealthyStatus health -> pure Available
```
Comment (KtorZ) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: We are doing this transformation in two places, which comfort me with the idea that `healthCheck` should really return a `HealthCheckSMASH` datatype in the first place instead of a `Maybe Bool`.

File: specifications/api/swagger.yaml [hunk 21]
```diff
@@ -4068,3 +4133,25 @@ paths:
...
+      tags: ["Utils"]
+      summary: SMASH health
+      description: |
+        Get health status of given SMASH uri.
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema: *ApiSmashServer
+      responses: *responsesPostSmashHealth
```
Comment (KtorZ) [inline specifications/api/swagger.yaml]: A `POST` request is awkward here, there's no resources being created. May we stick to a more restful approach and offer instead one single endpoint `GET /v2/smash/health` which also accepts an optional query parameter `url` or type `ApiSmashServer`; when provided, we'll fetch health from this URL, when omitted, we'll fetch health from the currently configured smash server if any.

Comment (hasufell) [reply specifications/api/swagger.yaml]: I was just trying to avoid url encoding as query parameter, since it's pesky.

Comment (KtorZ) [reply specifications/api/swagger.yaml]: I'd favor pesky url-encoded query parameter over awkward non restful post request :blush:

File: specifications/api/swagger.yaml (hunk 19 above)
Comment (hasufell) [reply specifications/api/swagger.yaml]: I'm also confused :sweat_smile: 
It's a nice example no?

Comment (KtorZ) [reply specifications/api/swagger.yaml]: It is, but what the heck has this addition to do with the PR :sweat_smile:  ... again, some merge resolution which didn't went well I assume.

File: lib/core/test/unit/Cardano/Wallet/Api/Malformed.hs [hunk 22]
```diff
@@ -1151,6 +1151,13 @@ instance Malformed (BodyParam ApiMaintenanceActionPostData) where
           )
         ]
 
+instance Malformed (BodyParam (ApiT SmashServer)) where
+    malformed = first (BodyParam . Aeson.encode) <$>
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Api/Malformed.hs]: Is it still relevant?

**Review by KtorZ — APPROVED** [review]

File: lib/core/test/unit/Cardano/Wallet/Api/Malformed.hs (hunk 22 above)
Comment (hasufell) [reply lib/core/test/unit/Cardano/Wallet/Api/Malformed.hs]: I guess we don't have query param specs yet:

```
    -- TODO
    -- Capture request query params as QueryParam
    --
    -- newtype QueryParam t = QueryParam Text
    --     deriving (Typeable)
    --
    -- type MkQueryRequest api :: *
    -- gEveryQueryParams :: Proxy api -> Request -> MkQueryRequest api
```

Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Api/Malformed.hs]: Indeed we do not, but the malformed instances are used only if the underlying type is used in the API. And now, there is no API call using `ApiT SmashServer` as a body parameter.

**Comment by hasufell** [comment]
@KtorZ I rebased, maybe have a quick look again. Not the first time I messed up a rebase...

**Review by KtorZ — APPROVED** [review]
bors merge
